### PR TITLE
OCLM-38 - Subscribe to snapshot or released versions

### PR DESCRIPTION
**Pull request by tmarzeion**
What is done here:
- column `release_version` added to `openconceptlab_update` table, it is used to check if latest OCL release is newer than one downloaded by OpenMRS OCL Client.
- `fetchUpdates()` method in `OclClient` is now renamed to `fetchSnapshotUpdates()`.
- `fetchSnapshotUpdates()` is called only when `isSubscribedToSnapshot()` in `Subscription` returns `true`.
- `fetchInitialUpdates()` is renamed to `fetchReleaseUpdates()`
- `fetchReleaseUpdates()` overloaded method is added, this method checks if last downloaded version is up to date with OCL release version. If yes, old `fetchReleaseUpdates()` is called.

I've added 3 test methods used to start update.
`startUpdate_shouldStartInitialUpdate()` - Starts initial update
`startUpdate_shouldStartReleaseUpdate()` - Starts followup update (only release version updates)
`startUpdate_shouldStartSnapshotUpdate()` - Starts followup update (with SNAPSHOT versions included)
These methods are `@Ignored` because tests are fetching data from original source, so it takes a lot of time (about 13 mins per test) to finish these tests. This issue will be fixed when some custom, small sized test-exclusive data will be created.

There is also `update_shouldUpdateLatestDownloadedRelease()` test method, which checks if new `String lastDownloadedRelease` field in `Update` object is successfully saved into DB.

**Summary:**
This code works like it worked before, until `boolean isFetchingSnapshotUpdates` in `Subscription` isn't changed to `false`, which is going to be implemented in new UI.
when `isFetchingSnapshotUpdates` is set to false, there is new logic triggered in `Updater`, which calls `OclClient` overloaded `fetchReleaseUpdates()` method that does update if new source version is released.

### CODE REVIEW

**Comment by tmarzeion** [comment]
@rkorytkowski 
Code is updated, 
I've added 3 test methods used to start update.
`startUpdate_shouldStartInitialUpdate()` - Starts initial update
`startUpdate_shouldStartReleaseUpdate()` - Starts followup update (only release version updates)
`startUpdate_shouldStartSnapshotUpdate()` - Starts followup update (with SNAPSHOT versions included)
These methods are `@Ignored` because tests are fetching data from original source, so it takes a lot of time (about 13 mins per test) to finish these tests. This issue will be fixed when some custom, small sized test-exclusive data will be created.

There is also `update_shouldUpdateLatestDownloadedRelease()` test method, which checks if new `String lastDownloadedRelease' field in `Update` object is successfully saved into DB.

**Summary:**
This code works like it worked before, until `boolean isFetchingSnapshotUpdates` in `Subscription` isn't changed to `false`, which is going to be implemented in new UI.
when `isFetchingSnapshotUpdates` is set to false, there is new logic triggered in `Updater`, which calls `OclClient` overloaded `fetchReleaseUpdates()` method that does update if new source version is released.


# TL;DR - SHOULD WORK

**Comment by rkorytkowski** [comment]
It's best you reapply the changes on top of what @AdamGrzybkowski did and address my review comment.

File: api/src/main/java/org/openmrs/module/openconceptlab/Import.java [hunk 1]
```diff
@@ -45,6 +45,10 @@
 	@Basic
 	@Column(name = "ocl_date_started")
 	private Date oclDateStarted;
+
+	@Basic
+	@Column(name = "last_downloaded_release")
+	private String lastDownloadedRelease;
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/openconceptlab/Import.java]: I asked to rename the variable to releaseVersion. As I understand it's not only downloaded, but release version imported for this import.

File: api/src/test/java/org/openmrs/module/openconceptlab/UpdateServiceTest.java [hunk 2]
```diff
@@ -136,12 +146,108 @@ public void saveSubscription_shouldSaveSubscription() throws Exception {
...
+
+		importService.saveSubscription(newSubscription);
+        Importer importer = new Importer();
+
+        File tempDir = File.createTempFile("ocl", "");
+        FileUtils.deleteQuietly(tempDir);
+        tempDir.deleteOnExit();
+        OclClient oclClient = new OclClient(tempDir.getAbsolutePath());
+
+        importer.setOclClient(oclClient);
```
Comment (rkorytkowski) [inline api/src/test/java/org/openmrs/module/openconceptlab/UpdateServiceTest.java]: In addition to those tests let's actually have a test, which mocks OclClient.

File: api/src/main/java/org/openmrs/module/openconceptlab/Import.java (hunk 1 above)
Comment (tmarzeion) [inline api/src/main/java/org/openmrs/module/openconceptlab/Import.java]: Renamed

File: api/src/main/java/org/openmrs/module/openconceptlab/client/OclClient.java [hunk 3]
```diff
@@ -125,7 +123,17 @@ public OclResponse fetchInitialUpdates(String url, String token) throws IOExcept
 		
 		return extractResponse(exportUrlGet);
     }
-	
+
+	public OclResponse fetchReleaseUpdates(String url, String token, String lastDownloadedReleaseVersion) throws IOException {
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/openconceptlab/client/OclClient.java]: lastDownloadedReleaseVersion -> lastReleaseVersion

File: api/src/main/java/org/openmrs/module/openconceptlab/ImportService.java [hunk 4]
```diff
@@ -59,6 +59,8 @@
     @Transactional
 	void updateOclDateStarted(Import update, Date oclDateStarted);
 
+	void updateLastRelease(Import anImport, String version);
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/openconceptlab/ImportService.java]: Didn't I say to change the name to updateReleaseVersion?

**Comment by rkorytkowski** [comment]
@tmarzeion, are you having hard time following new github's review?

**Comment by tmarzeion** [comment]
@rkorytkowski There is something wrong with this review, since I didn't see any comments before (for example, first review was requested yesterday at 11:22, but comments for this review were created today about 11:45)
